server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java

```diff
@@ -0,0 +1,208 @@
+    } catch (MutationsRejectedException mex) {
+      log.warn("Failed to update reference for table: " + tableName);
+      log.warn("Constraint violations: {}", mex.getConstraintViolationSummaries());
+      throw new IllegalStateException("Failed to process table: " + tableName, mex);
+    } catch (Exception ex) {
+      throw new IllegalStateException("Failed to process table: " + tableName, ex);
+    }
+  }
+
+  @VisibleForTesting
+  void upgradeDataFileCF(final Key key, final Value value, final BatchWriter batchWriter,
+      final String tableName) {
+    String file = key.getColumnQualifier().toString();
+    // filter out references if they are in the correct format already.
+    if (fileNeedsConversion(file)) {
+      var fileJson = StoredTabletFile.of(new Path(file)).getMetadataText();
+      try {
+        Mutation update = new Mutation(key.getRow());
+        update.at().family(DataFileColumnFamily.STR_NAME).qualifier(fileJson).put(value);
+        log.trace("table: {}, adding: {}", tableName, update.prettyPrint());
+        batchWriter.addMutation(update);
+
+        Mutation delete = new Mutation(key.getRow());
+        delete.at().family(DataFileColumnFamily.STR_NAME).qualifier(file).delete();
+        log.trace("table {}: deleting: {}", tableName, delete.prettyPrint());
+        batchWriter.addMutation(delete);
+      } catch (MutationsRejectedException ex) {
+        // include constraint violation info in log - but stop upgrade
+        log.warn(
+            "Failed to update file reference for table: " + tableName + ". row: " + key.getRow());
+        log.warn("Constraint violations: {}", ex.getConstraintViolationSummaries());
+        throw new IllegalStateException("File conversion failed. Aborting upgrade", ex);
+      }
+    }
+  }
+
+  @VisibleForTesting
+  void removeChoppedCF(final Key key, final BatchWriter batchWriter, final String tableName) {
+    Mutation delete = null;
+    try {
+      delete = new Mutation(key.getRow()).at().family(ChoppedColumnFamily.STR_NAME)
+          .qualifier(ChoppedColumnFamily.STR_NAME).delete();
+      log.warn(
+          "Deleting chopped reference from:{}. Previous split or delete may not have completed cleanly. Ref: {}",
+          tableName, delete.prettyPrint());
+      batchWriter.addMutation(delete);
+    } catch (MutationsRejectedException ex) {
+      log.warn("Failed to delete obsolete chopped CF reference for table: " + tableName + ". Ref: "
+          + delete.prettyPrint() + ". Will try to continue. Ref may need to be manually removed");
+      log.warn("Constraint violations: {}", ex.getConstraintViolationSummaries());
+    }
+  }
+
+  @VisibleForTesting
+  void removeExternalCompactionCF(final Key key, final BatchWriter batchWriter,
+      final String tableName) {
+    Mutation delete = null;
+    try {
+      delete = new Mutation(key.getRow()).at().family(ExternalCompactionColumnFamily.NAME)
+          .qualifier(key.getColumnQualifier()).delete();
+      log.warn(
```

Comment:
The warn was so that that would stand out in the logs.  This will run once, and debug statements could be easier lost / ignored.  The removal of external compaction markers is because they would likely cause problems because they do not include the fence parameters.

Reply:
Its not expected that an external compaction would run across an upgrade and removing entries related to any that were running is not very interesting.  It could possibly log a lot of warns if there are a lot of ecomp entries and that would be a lot of noise that is not very useful IMO.

Reply:
Currently, we block the upgrade from even starting if there are outstanding FATE transactions - do we need something equivalent to stop / block external compactions?

Also, during the upgrade only the root and the metadata tablets would be online and I don't think we would want concurrent operations occurring on those tables while the upgrade is executing.

My assumption was that because no FATEs are present, any flags present would be from operations that terminated and did not get a chance to clean-up.  I may be important to know which things did not stop cleanly - so it can be determined if post-upgrade clean-up operations, such as re-queuing the interrupted compaction is necessary.

Reply:
Changed to debug in 77dd24939a